trackintel/preprocessing/triplegs.py

```diff
@@ -145,31 +149,40 @@ def generate_trips(spts, tpls, gap_threshold=15):
     spts_tpls_no_act = spts_tpls[~spts_tpls["activity"]]
     spts_tpls_only_act = spts_tpls[spts_tpls["activity"]]
 
+    # todo: maybe move this part down to where you need it
     # add gaps as activities, to simplify id assignment later.
     gaps = pd.DataFrame(spts_tpls.loc[gap, "user_id"])
     gaps["started_at"] = spts_tpls.loc[gap, "finished_at"] + gap_threshold / 2
     gaps[["type", "activity", "new_trip"]] = ["gap", True, True]
 
     # same for user changes
     user_change = pd.DataFrame(spts_tpls.loc[condition_new_user, "user_id"])
+    # todo: if the started_at will be used for sorting then the time difference should be as small as possible
     user_change["started_at"] = spts_tpls.loc[condition_new_user, "started_at"] - gap_threshold / 2
     user_change[["type", "activity", "new_trip"]] = ["user_change", True, True]
 
     trips_grouper = spts_tpls_no_act.groupby("trip_id")
+    # todo: for me using `np.array` raised an exception: "ValueError: Must produce aggregated value"
     trips = trips_grouper.agg(
-        {"user_id": "mean", "started_at": min, "finished_at": max, "type": np.array, "id": np.array,}
+        {"user_id": "mean", "started_at": min, "finished_at": max, "type":list, "id": list,}
     )
 
     def _seperate_ids(row):
         """Split aggregated ids into staypoints and triplegs arrays."""
-        t = row["type"] == "tripleg"
-        tpls_ids = row.loc["id"][t]
-        spts_ids = row.loc["id"][np.logical_not(t)]
+        row_type = np.array(row["type"])
+        row_id = np.array(row["id"])
+        t = row_type == "tripleg"
+        tpls_ids = row_id[t]
+        spts_ids = row_id[np.logical_not(t)]
         return [spts_ids, tpls_ids]
 
     trips[["spts", "tpls"]] = trips.apply(_seperate_ids, axis=1, result_type="expand")
 
     # save trip_id as column
+    # todo: I don't understand this. The trip id is the index. Therefore the reset operation creates the column "trip
```

Comment:
The groupby deletes empty trips, so we may have fewer trips than originally planned. The reason for this is not quite clear to me so far. So this reset_index is here to make sure that our trip index really goes from 0 to n-1. I'll have a look into it.

Reply:
So for example I've we have `activity - gap - tripleg` we have an empty trip as the next trip already starts on the tripleg so the activity has no "non" activities we can aggregate &rarr; therefor we need to reindex to get the trip id without the empty trips.

Reply:
👍